src/substituter.ts

```diff
@@ -0,0 +1,106 @@
+    if (left.type === 'Literal') {
+      if (right.type === 'Literal') {
+        const error = rttc.checkBinaryExpression(node, operator, left.value, right.value)
+        if (error === undefined) {
+          return [ast.literal(evaluateBinaryExpression(operator, left.value, right.value)), context]
+        } else {
+          throw error
+        }
+      } else {
+        const [reducedRight] = reduce(right, context)
+        const reducedExpression = ast.binaryExpression(
+          operator,
+          left,
+          reducedRight as es.Expression,
+          node.loc!
+        )
+        return [reducedExpression, context]
+      }
+    } else {
+      const [reducedLeft] = reduce(node.left, context)
+      const reducedExpression = ast.binaryExpression(
+        operator,
+        reducedLeft as es.Expression,
+        right,
+        node.loc!
+      )
+      return [reducedExpression, context]
+    }
+  },
+  Program(node: es.Program, context: Context) {
+    const [firstStatement, ...otherStatements] = node.body
+    if (
+      firstStatement.type === 'ExpressionStatement' &&
+      firstStatement.expression.type === 'Literal'
+    ) {
+      return [ast.program(otherStatements, node.loc!), context]
+    } else {
+      const [reduced, newContext] = reduce(firstStatement, context)
+      return [ast.program([reduced as es.Statement, ...otherStatements], node.loc!), newContext]
+    }
+  },
+  /* <BAD PRACTICE>
+   * for now BlockStatement is just duplicated code from Program which is BAD PRACTICE,
+   * but in future BlockStatement will need to handle return as well.
+   * </BAD PRACTICE>
+   */
+  BlockStatement(node: es.BlockStatement, context: Context) {
+    const [firstStatement, ...otherStatements] = node.body
+    const isFirstStatementCompletelyReduced =
+      firstStatement.type === 'ExpressionStatement' && firstStatement.expression.type === 'Literal'
+    if (isFirstStatementCompletelyReduced) {
+      return [ast.blockStatement(otherStatements, node.loc!), context]
+    } else {
+      const [reduced, newContext] = reduce(firstStatement, context)
+      return [
+        ast.blockStatement([reduced as es.Statement, ...otherStatements], node.loc!),
+        newContext
+      ]
+    }
+  }
+}
+
+function reduce(node: es.Node, context: Context): [es.Node, Context] {
+  const reducer = reducers[node.type]
+  if (reducer === undefined) {
+    return [node, context] // if reducer is not found we just get stuck
+  } else {
+    return reducer(node, context)
+  }
+}
+
+export function getEvaluationSteps(code: string, context: Context): Array<[es.Node, Context]> {
+  const steps: Array<[es.Node, Context]> = []
+  try {
+    const program = parse(code, context)
+    if (program === undefined) {
+      return [[parse('', context)!, context]]
+    }
+    steps.push([program, context])
+    let [reduced, newContext] = reduce(program, context)
+    while ((reduced as es.Program).body.length > 0) {
```

Comment:
I think the logic for whether the reduction has ended should be something else... I would have Program/Block reduce to a literal and then for literals to stop the reduction.

Tell me what you think is a better idea.

Reply:
I think I'll make a thunk class that has the property isReducible.

Reply:
We kind of agreed that the result of an evaluation should be an array of programs (syntax trees). Each element of the array is a step in the small-step reduction. I agree with a "reducible" property, but not with raising an exception in case a program becomes irreducible. The frontend can then decide how to handle. There may be a way to highlight the issue that leads to irreducibility (type error, undeclared name, etc).

Reply:
Errors such as these will be caught (currently string * number already throws an error), so there shouldn't be a case of us not knowing what to do.

Reply:
Also, the current throwing of an Exception of irreducible statements is a temporary measure since not everything is implemented.

Reply:
Maybe it's good to throw an exception, and handle it, and as usual show an error in the REPL. The SUBST Inspector window can still show the sequence of small-steps.